### PR TITLE
Kl/l2 bridge fixes

### PR DESCRIPTION
**Pull request by kelemeno**
# What ❔

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

File: l2-contracts/contracts/bridge/L2AssetRouter.sol [hunk 1]
```diff
@@ -22,13 +21,16 @@ import {EmptyAddress, InvalidCaller} from "../L2ContractErrors.sol";
...
+contract L2AssetRouter is IL2AssetRouter, Initializable {
     /// @dev Chain ID of Era for legacy reasons
     uint256 public immutable ERA_CHAIN_ID;
 
     /// @dev Chain ID of L1 for bridging reasons
     uint256 public immutable L1_CHAIN_ID;
 
+    /// @dev The address of the L2 legacy shared bridge.
+    address public L2_LEGACY_SHARED_BRIDGE;
+
```
Comment (StanislavBreadless) [inline l2-contracts/contracts/bridge/L2AssetRouter.sol:33]: not from this line, but could you please remove the `DEPRECATED` fields from the L2AssetRouter if is not the shared bridge anymore?

File: l2-contracts/contracts/bridge/L2AssetRouter.sol [hunk 2]
```diff
@@ -66,13 +68,28 @@ contract L2AssetRouter is IL2AssetRouter, ILegacyL2SharedBridge, Initializable {
         _;
     }
 
-    /// @dev Contract is expected to be used as proxy implementation.
+    /// @notice Checks that the message sender is the legacy L2 bridge.
+    modifier onlyLegacyBridge() {
+        // Only the L1 bridge counterpart can initiate and finalize the deposit.
```
Comment (StanislavBreadless) [inline l2-contracts/contracts/bridge/L2AssetRouter.sol:73]: the comment is incorrect

File: l2-contracts/contracts/bridge/L2AssetRouter.sol (hunk 1 above)
Comment (StanislavBreadless) [reply l2-contracts/contracts/bridge/L2AssetRouter.sol:33]: also `l1TokenAddress` is not needed there

File: l2-contracts/contracts/bridge/L2SharedBridgeLegacy.sol [hunk 3]
```diff
@@ -2,26 +2,142 @@
...
+        if (token != address(0)) {
+            return token;
+        }
+        return calculateCreate2TokenAddress(_l1Token);
+    }
+
+    /// @notice Calculates L2 wrapped token address given the currently stored beacon proxy bytecode hash and beacon address.
+    /// @param _l1Token The address of token on L1.
+    /// @return Address of an L2 token counterpart.
+    function calculateCreate2TokenAddress(address _l1Token) internal view returns (address) {
```
Comment (StanislavBreadless) [inline l2-contracts/contracts/bridge/L2SharedBridgeLegacy.sol:111]: internal methods should start with an underscore

File: l2-contracts/contracts/bridge/L2SharedBridgeLegacy.sol [hunk 4]
```diff
@@ -2,26 +2,142 @@
...
+
+    /// @dev Convert the L1 token address to the create2 salt of deployed L2 token
+    function _getCreate2Salt(address _l1Token) internal pure returns (bytes32 salt) {
+        salt = bytes32(uint256(uint160(_l1Token)));
+    }
+
+    /// @dev Deploy the beacon proxy for the L2 token, while using ContractDeployer system contract.
+    /// @dev This function uses raw call to ContractDeployer to make sure that exactly `l2TokenProxyBytecodeHash` is used
+    /// for the code of the proxy.
+    function deployBeaconProxy(bytes32 salt) external returns (address proxy) {
```
Comment (StanislavBreadless) [inline l2-contracts/contracts/bridge/L2SharedBridgeLegacy.sol:126]: please add access control. someone from outside could deploy a beacon proxy with a salt equal to an L1 token address and make all deployments coming from NTV fail

File: l2-contracts/contracts/bridge/L2NativeTokenVault.sol [hunk 5]
```diff
@@ -87,9 +89,16 @@ contract L2NativeTokenVault is IL2NativeTokenVault, Ownable2StepUpgradeable {
             address l2StandardToken = address(new L2StandardERC20{salt: bytes32(0)}());
             l2TokenBeacon = new UpgradeableBeacon{salt: bytes32(0)}(l2StandardToken);
             l2TokenBeacon.transferOwnership(owner());
+            emit L2TokenBeaconUpdated(_l2TokenBeacon, _l2TokenProxyBytecodeHash);
         }
     }
 
+    function setLegacyTokenAssetId(address _l2TokenAddress) public {
```
[the remark this answers is not in the document]
Comment (kelemeno) [reply l2-contracts/contracts/bridge/L2NativeTokenVault.sol:96]: I think it will be useful for migration

**Pull request merged**